### DISCUSSION TITLE
Using After trigger with DRF APITestCase

### DISCUSSION THREAD

**Discussion by melvinkcx** [AmbitionEng/django-pgtrigger#197] [Q&A]
Hi, first of all, thank you for this awesome library! However, I came across a petite problem and I wasn't sure if I was using the library correctly. 

I wanted to create a trigger that runs after an insertion:

```python
pgtrigger.Trigger(
            name="<name_1>",
            when=pgtrigger.After,
            operation=pgtrigger.Insert,
            timing=pgtrigger.Immediate,
            func=f"""
                NEW.col_1 = 'A GENERATED VALUE';
                RETURN NEW;
            """
        )
```

The trigger worked when I tried to insert data manually, either directly into the table through psql or Django models. However, it didn't work in DRF APITestCase.

By raising an exception in the trigger's `func`, I could see that the trigger wasn't trigger until `tearDown()` was called. 

I was wondering if I was using the trigger correctly or if it's a known issue. I couldn't find any relevant Github issues.

In my case, I eventually changed the logic of my trigger and used `when=pgtrigger.Before`.

I appreciate it if anyone here can confirm if this is a known issue. Thank you

**Comment by wesleykendall** [comment]
How did you verify it didn't work?

A few things to keep in mind:

- If you test a trigger in a test case, be sure that you refresh the model instance. Just like [Django generated fields](https://docs.djangoproject.com/en/5.1/ref/models/fields/#generatedfield), calling `.save()` on a model with a trigger doesn't update the instance. You have to call `.refresh_from_db()`. Did you forget to call this before asserting that your `col_1` was generated?
- `timing=pgtrigger.Immediate` isn't needed here, that's the default value. However, what caused you to override this in the first place? If you declared this as a [deferrable trigger](https://django-pgtrigger.readthedocs.io/en/4.13.3/deferrable/), it wouldn't execute until the end of the transaction, which may explain the behavior you were seeing since I assume APITestCase is going to wrap everything in a transaction
- Are your migrations up to date and were they applied properly when running this test case?

**Reply by wesleykendall** [reply]
FYI @melvinkcx I converted this to a discussion for now as this is very likely not an issue in the library. Let me know if my answer solves your question

**Reply by melvinkcx** [reply]
Hi, thanks for your reply. To answer your questions, 

1. Yes, in my test cases, I re-fetched the record using `MyModel.objects.get()`, so it should have the latest data.
2. The `timing=pgtrigger.Immediate` was to make it explicit so that my future self won't have to guess or figure out what the default value was
3. Yes. I even added a breakpoint to inspect `pgtrigger.registered` and the trigger was indeed there.

**Reply by melvinkcx** [reply]
I brought this up just to know if I was doing it correctly, it is entirely possible that the issue has something to do with my test setup.
